segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageProvider.java

```diff
@@ -0,0 +1,179 @@
+     * @return int Number of bytes read.
+     * @throws IOException Throws IOException in case of I/O related exceptions.
+     * @throws IllegalArgumentException If argument is invalid.
+     * @throws NullPointerException  If the parameter is null.
+     * @throws IndexOutOfBoundsException If the index is out of bounds.
+     */
+    int read(ChunkHandle handle, long fromOffset, int length, byte[] buffer, int bufferOffset) throws IOException, NullPointerException, IndexOutOfBoundsException;
+
+
+    /**
+     * Writes the given data to the underlying storage object.
+     *
+     * @param handle ChunkHandle of the storage object to write to.
+     * @param offset Offset in the file to start writing.
+     * @param length Number of bytes to write.
+     * @param data An InputStream representing the data to write.
+     * @return int Number of bytes written.
+     * @throws IOException Throws IOException in case of I/O related exceptions.
+     */
+    int write(ChunkHandle handle, long offset, int length, InputStream data) throws IOException;
+
+    /**
+     * Concatenates two or more chunks using native facility. The first chunk is concatenated to.
+     *
+     * @param chunks Array of ChunkHandle to existing chunks to be appended together. The chunks are appended in the same sequence the names are provided.
+     * @param lengths Lengths of chunks.
+     * @return int Number of bytes concatenated.
+     * @throws IOException Throws IOException in case of I/O related exceptions.
+     * @throws UnsupportedOperationException If this operation is not supported by this provider.
+     */
+    int concatNatively(ChunkHandle[] chunks, long[] lengths) throws IOException, UnsupportedOperationException;
+
+    /**
+     * Concatenates two or more chunks using append operation. The first chunk is concatenated to.
+     *
+     * @param chunks Array of ChunkHandle to existing chunks to be appended together. The chunks are appended in the same sequence the names are provided.
+     * @param lengths Lengths of chunks.
+     * @return int Number of bytes concatenated.
+     * @throws IOException Throws IOException in case of I/O related exceptions.
+     * @throws UnsupportedOperationException If this operation is not supported by this provider.
+     */
+    int concatWithAppend(ChunkHandle[] chunks, long[] lengths) throws IOException, UnsupportedOperationException;
+
+    /**
+     * Truncates a given chunk.
+     *
+     * @param handle ChunkHandle of the storage object to truncate.
+     * @param offset Offset to truncate to.
+     * @return True if the object was truncated, false otherwise.
+     * @throws IOException Throws IOException in case of I/O related exceptions.
+     * @throws UnsupportedOperationException If this operation is not supported by this provider.
+     */
+    boolean truncate(ChunkHandle handle, long offset) throws IOException, UnsupportedOperationException;
+
+    /**
+     * Sets readonly attribute for the chunk.
```

Comment:
This is the first time we are talking about attributes for chunks in this class, which makes me wonder whether we should be leaking this concept of attributes for chunks at all.

Reply:
This method allows to mark chunk as readonly. It is useful to make chunks readonly when you are done writing to them. But not absolutely essential either.

Reply:
I makes sense, the confusing part is mentioning attributes, which is not a concept that has appeared in this interface until here. It makes it feel like there are other attributes, and we don't say what they are here.

Reply:
removed word attribute